eth_keys/datatypes.py

```diff
@@ -57,163 +58,195 @@
         from eth_keys.backends import get_backend
-
-        if cls._backend is None:
-            return get_backend()
-        else:
-            return cls._backend
+        return get_backend(*args, **kwargs)
 
 
 class BaseKey(ByteString, collections.Hashable):
     _raw_key = None  # type: bytes
 
-    def to_hex(self):
-        # type: () -> Text
+    def to_hex(self) -> str:
         # Need the 'type: ignore' comment below because of
         # https://github.com/python/typeshed/issues/300
         return '0x' + codecs.decode(codecs.encode(self._raw_key, 'hex'), 'ascii')  # type: ignore
 
-    def to_bytes(self):
-        # type: () -> bytes
+    def to_bytes(self) -> bytes:
         return self._raw_key
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         return big_endian_to_int(keccak(self.to_bytes()))
 
-    def __str__(self):
+    def __str__(self) -> str:
         return self.to_hex()
 
-    def __unicode__(self):
-        return self.__str__()
-
-    def __int__(self):
+    def __int__(self) -> int:
         return big_endian_to_int(self._raw_key)
 
-    def __len__(self):
+    def __len__(self) -> int:
+        # TODO: this seems wrong.
         return 64
 
-    def __getitem__(self, index):
+    # Must be typed with `ignore` due to
+    # https://github.com/python/mypy/issues/1237
+    def __getitem__(self, index: int) -> int:  # type: ignore
         return self._raw_key[index]
 
-    def __eq__(self,
-               other  # type: Any
-               ):
-        # type: (...) -> bool
+    def __eq__(self, other: Any) -> bool:
         if hasattr(other, 'to_bytes'):
             return self.to_bytes() == other.to_bytes()
         elif is_bytes(other):
             return self.to_bytes() == other
         else:
             return False
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return "'{0}'".format(self.to_hex())
 
-    def __index__(self):
+    def __index__(self) -> int:
         return self.__int__()
 
-    def __hex__(self):
-        # type: () -> Text
+    def __hex__(self) -> str:
         if sys.version_info[0] == 2:
             return codecs.encode(self.to_hex(), 'ascii')
         else:
             return self.to_hex()
 
 
-class PublicKey(BaseKey, BackendProxied):
-    def __init__(self, public_key_bytes):
-        # type: (bytes) -> None
+class PublicKey(BaseKey, LazyBackend):
+    def __init__(self, public_key_bytes: bytes, **kwargs) -> None:
         validate_public_key_bytes(public_key_bytes)
 
         self._raw_key = public_key_bytes
+        super().__init__(**kwargs)
 
     @classmethod
-    def from_private(cls, private_key):
-        # type: (PrivateKey) -> PublicKey
-        return cls.get_backend().private_key_to_public_key(private_key)
+    def from_private(cls,
+                     private_key,  # type: PrivateKey
+                     backend=None,  # type: BaseECCBackend
+                     ):
+        # type: (...) -> PublicKey
```

Comment:
Should we convert these to py3-style as well?

Reply:
Same issue here.  `PrivateKey` is not yet defined at this point in the module so doing python3 style caused syntax errors.

Reply:
In those cases you can enclose the name in single quotes